allennlp/models/semantic_parsing/drop/naqanet.py

```diff
@@ -152,28 +189,100 @@ def forward(self,  # type: ignore
         passage_vector = util.weighted_sum(modeled_passage_list[0], passage_weights)
         # The vector representation of question is calculated based on the unmatched encoding,
         # because we may want to infer the answer ability only based on the question words.
+
         question_weights = self._question_weights_predictor(encoded_question).squeeze(-1)
         question_weights = masked_softmax(question_weights, question_mask)
         question_vector = util.weighted_sum(encoded_question, question_weights)
 
+        worlds = [[DropNaqanetLanguage(encoded_question=encoded_question[i].unsqueeze(0), 
+                                       question_mask=question_mask[i].unsqueeze(0),
+                                       passage_vector=passage_vector[i].unsqueeze(0),
+                                       passage_mask=passage_mask[i].unsqueeze(0),
+                                       modeled_passage_list=[passage[i].unsqueeze(0) for passage in modeled_passage_list],
+                                       number_indices=number_indices[0].unsqueeze(0), 
+                                       parameters=self.naqanet_parameters)] for i in range(batch_size)]
+        
+        initial_rnn_state = [] 
+        encoded_question_list = [encoded_question[i] for i in range(batch_size)]
+        question_mask_list = [question_mask[i] for i in range(batch_size)]
+        memory_cell = encoded_question.new_zeros(batch_size, self.encoding_out_dim) 
+
+        final_encoder_output = util.get_final_encoder_states(encoded_question,
+                                                             question_mask)
+        
         for i in range(batch_size):
-            # TODO(mattg): split the state, make a World and an initial RnnStatelet and
-            # GrammarStatelet, and then do a search.
-            pass
-        # TODO(mattg): Construct a semantic parser here, and run a search.  Take the result of the
-        # search, multiply the program probability by the answer probability, then sum across all
-        # programs.
-
-        # If answer is given, compute the loss.
-        # if answer_as_passage_spans is not None or answer_as_question_spans is not None \
-                # or answer_as_add_sub_expressions is not None or answer_as_counts is not None:
-            # all_log_marginal_likelihoods = torch.stack(log_marginal_likelihood_list, dim=-1)
-            # all_log_marginal_likelihoods = all_log_marginal_likelihoods + answer_ability_log_probs
-            # marginal_log_likelihood = util.logsumexp(all_log_marginal_likelihoods)
+            initial_rnn_state.append(RnnStatelet(hidden_state=final_encoder_output[i],
```

Comment:
Make the initial hidden state be a `concat([passage_vector, question_vector])`.  The way you've done this here, you're not using the passage as an input to which program you should pick, whereas NAQANet does use this.